src/cli/self_update.rs

```diff
@@ -746,21 +748,43 @@ fn maybe_install_rust(
         info!("default host triple is {}", cfg.get_default_host_triple()?);
     }
 
-    // If there is already an install, then `toolchain_str` may not be
-    // a toolchain the user actually wants. Don't do anything.  FIXME:
-    // This logic should be part of InstallOpts so that it isn't
-    // possible to select a toolchain then have it not be installed.
-    if toolchain_str == "none" {
+    let user_specified_something =
+        toolchain.is_some() || !targets.is_empty() || !components.is_empty();
+
+    // If the user specified they want no toolchain, we skip this, otherwise
+    // if they specify something directly, or we have no default, then we install
+    // a toolchain (updating if it's already present) and then if neither of
+    // those are true, we have a user who doesn't mind, and already has an
+    // install, so we leave their setup alone.
+    if toolchain == Some("none") {
         info!("skipping toolchain installation");
+        if !components.is_empty() {
+            warn!(
+                "ignoring requested component{}: {}",
+                if components.len() == 1 { "" } else { "s" },
+                components.join(", ")
+            );
+        }
+        if !targets.is_empty() {
+            warn!(
+                "ignoring requested target{}: {}",
+                if targets.len() == 1 { "" } else { "s" },
+                targets.join(", ")
+            );
+        }
         println!();
-    } else if cfg.find_default()?.is_none() {
+    } else if user_specified_something || cfg.find_default()?.is_none() {
+        let toolchain_str = toolchain.unwrap_or("stable");
         let toolchain = cfg.get_toolchain(toolchain_str, false)?;
+        if toolchain.exists() {
+            warn!("Updating existing toolchain, profile choice will be ignored");
```

Comment:
Is this missing logic, or not well defined?

Reply:
At the moment it's not clear what it'd mean - should it only *add* missing components?  Should it make it *be* only the components for the profile (e.g. if you say `minimal` should it remove anything else?) - as such, this is punted just like it is if you do `rustup toolchain install ...` after it's already installed, the profile will be ignored.

Reply:
`rustup set` says "The default components installed" about profile.
`rustup component` says we can add and remove components

Based on that I would assume that the profile only determines what is added on install. If the toolchain already exists I would expect it to add missing components and leave the rest alone.

Since the rust-init option is called `--profile` (no default) I would also assume it only affects that particular call. No idea if that is the current behavior.

Reply:
So your proposal would be "If the user explicitly invoked `rustup toolchain install` (or `rustup-init` as is considered in this PR) then any components missing from the profile of an already installed toolchain should explicitly be added in.  I *think* that could be consistent but wouldn't be behaviour belonging in *this* PR.  I'd be up for an issue about that being opened though, if either of you agree with my restating of things.

Reply:
I'm happy with 'this is not well defined yet, lets open a new issue to discuss'.

That proposed behaviour might be it, but another one might be to coerce back to the profiles exact definition - lets work through it elsewhere.